Documentation/Registry.md

```diff
@@ -661,23 +661,37 @@ as described by [RFC 6249].
 A client MAY use this information
 to determine its preferred strategy for downloading.
 
-A server that indexes but doesn't host packages
-MAY respond with a status code of `303` (See Other)
-and redirect to a hosted package archive if one is available.
+A server MAY respond with a status code of `303` (See Other)
+to redirect the client to download the source archive from another host.
+The client MUST NOT follow redirects that downgrade to an insecure connection.
+The client SHOULD limit the number of redirects to prevent a redirect loop.
+
+For example,
+a server redirects the client to download from
+a content delivery network (CDN) using a signed URL:
 
 ```http
 HTTP/1.1 303 See Other
-Content-Version: 1
+Location: https://example.cdn.com/LinkedList-1.1.1.zip?key=XXXXXXXXXXXXXXXXX
+```
+
+```http
+GET /LinkedList-1.1.1.zip?key=XXXXXXXXXXXXXXXXX HTTP/1.1
+Host: example.cdn.com
+Accept: application/vnd.swift.registry.v1+zip
+```
+
+```http
+HTTP/1.1 200 OK
+Accept-Ranges: bytes
+Cache-Control: public, immutable
+Content-Type: application/zip
+Content-Disposition: attachment; filename="LinkedList-1.1.1.zip"
 Content-Length: 2048
+Content-Version: 1
 Digest: sha-256=a2ac54cf25fbc1ad0028f03f0aa4b96833b83bb05a14e510892bb27dea4dc812
```

Comment:
@mattt Diffing the latest changes against https://github.com/apple/swift-package-manager/pull/3624/commits/5e2b7acac36b4ff859adae00f768c6e5042ac974, are we going to keep `Digest` but remove `Location`?

Reply:
This example shows a sequence of:

1. A redirect response from the registry
2. A new request from the client using the `Location` of the redirect
3. A response from the new, redirected host

The first response includes a `Location` because it's a redirect, but the second response doesn't because it's serving the asset.

Reply:
Thanks, so this is proposing to remove `Digest` from the registry's redirect response.

There was a discussion on this in the past, and having `Digest` in the redirect response addresses a concern that @robertlacroix had: 

Screenshot [https://forums.swift.org/t/se-0292-3rd-review-package-registry-service/49107/5](https://forums.swift.org/t/se-0292-3rd-review-package-registry-service/49107/5):
<img width="767" alt="image" src="https://user-images.githubusercontent.com/12647725/127211557-bf021bf7-6b0a-46be-ae92-d4d79180b0ec.png">



Screenshot [https://forums.swift.org/t/se-0292-3rd-review-package-registry-service/49107/7](https://forums.swift.org/t/se-0292-3rd-review-package-registry-service/49107/7):
<img width="770" alt="image" src="https://user-images.githubusercontent.com/12647725/127211400-dc2486cc-e795-46a8-835b-cfe2f3d3a82b.png">

I want to get @robertlacroix's feedback here to make sure he is ok with this change.

Reply:
Copying my rationale from https://github.com/apple/swift-evolution/pull/1410 , for convenience: 

> **Redirection**: The current guidance to treat forwarding headers as authoritative is (as far as I can tell) unprecedented and possibly out-of-spec for HTTP. Our understanding of the feedback in https://github.com/apple/swift-evolution/pull/1319 is that the motivation for this change was to ensure that proxying services are supported. Conventional HTTP redirects support this use case without any special accommodations.

Reply:
We added the digest to the package metadata returned by the registry, right? Therefore I'd prefer to remove examples of `Digest` headers from the responses serving the archive for two reasons:

1) There's a single source of truth (package metadata) and to be clear that clients must always use that as opposed to any `Digest` header
2) The archive might be served from a CDN that doesn't support that header

Reply:
> We added the digest to the package metadata returned by the registry, right? 

That was an idea that was suggested during 3rd review, but didn't make it into the final proposal.

> Therefore I'd prefer to remove examples of `Digest` headers from the responses serving the archive for two reasons:
> 
> 1. There's a single source of truth (package metadata) and to be clear that clients must always use that as opposed to any `Digest` header

I agree that a single source of truth is preferable for client validation. Our design is for the `Digest` header to serve that role. An advantage to this approach is that not additional requests are needed to get the digest when downloading a source archive for a resolved package version.

> 2. The archive might be served from a CDN that doesn't support that header

I'm unaware of any CDN providers that don't support origin header. I've personally worked with AWS CloudFront, Azure, Cloudflare, and Fastly, and know for a fact that they all support sending custom response headers. I'm not aware of any that do this automatically based on content, but you can at least set a precomputed header on the resource. Skimming the docs of other providers, it looks like Alibaba, Akamai, and Google Cloud also support custom headers.

Reply:
I would really like the registry to be in control of the authenticity of the package. As it's designed right now, a redirect on the `/{scope}/{name}/{version}.zip` endpoint yields control of the checksum to another server, which adds opportunity to maliciously alter the package.

Therefore I propose adding the digest to `/{scope}/{name}` and be that the source of truth. I agree it requires another request in some cases.

Reply:
seems like a reasonable request to add this to the versions list API which SwiftPM hits anyways as part of the dependency resolution / version determination flow.

Reply:
@robertlacroix Thanks for weighing in on this. I just pushed 362099ae, which updates the response for `GET /{scope}/{name}/{version}` to include a `checksum` field that serves as the authoritative integrity check for package resources. This change also provides a more predictable, extensible structure to `GET /{scope}/{name}/{version}`. It also fixes a mistake that I didn't realize until now, which is that the HTTP `Digest` field defines `sha-256` to use a Base64 encoding rather than the hexadecimal representation used by `swift package generate-checksum`.

Reply:
Should `checksum` be added to both `GET /{scope}/{name}` and `GET /{scope}/{name}/{version}`? SwiftPM will hit the list package releases API more frequently I would think and having `checksum` in that response should serve as an optimization.

Reply:
@yim-lee I think the benefit of having a single source of truth outweighs the potential drawback of an extra HTTP request. And if we ever expand what resources are available (for binary distribution(s), software bill of materials, documentation archive, etc.), it'll be better to have those all together under `GET /{scope}/{name}/{version}`, rather than congesting the list of versions in `GET /{scope}/{name}`.